cibyl/sources/zuuld/git_api.py

```diff
@@ -0,0 +1,216 @@
+import base64
+import json
+import logging
+import os
+import urllib
+
+import requests
+import yaml
+
+from cibyl.exceptions.zuul import RateLimitException
+from cibyl.models.attribute import AttributeDictValue
+from cibyl.models.ci.job import Job
+from cibyl.models.ci.tenant import Tenant
+from cibyl.sources.git import GitSource
+from cibyl.utils.yaml import get_loader
+
+ZUUL_JOBS_LOCATION = ['zuul.d', '.zuul.yaml']
+LOG = logging.getLogger(__name__)
+
+
+class ZuulGitHub:
+    def __init__(self, kwargs=None):
+        if kwargs is None:
+            kwargs = {}
+        self.git_base_url = "https://api.github.com/repos/{}/git/trees/master"
+        self.job_dir = "zuul.d"
+        self.username = kwargs.get("username", "")
+        self.token = kwargs.get("token", "")
+        self.repos = kwargs.get('repos', None)
+        self._validate()
+
+        # self.req_url = self.git_base_url.format(self.path)
+
+    def _validate(self):
+        # validate url
+        for repo in self.repos:
+            try:
+                result = urllib.parse.urlparse(repo.get('url'))
+                self.scheme = result.scheme
+                self.netloc = result.netloc
+                if result.path.endswith(".git"):
+                    temp_path = result.path.split(".git")[0]
+                    if temp_path.startswith("/"):
+                        temp_path = temp_path.lstrip("/")
+                else:
+                    if result.path.startswith("/"):
+                        temp_path = result.path.lstrip("/")
+
+                self.path = repo['repo_name'] = temp_path
+                self.params = result.params
+                self.query = result.query
+            except Exception as e:
+                raise urllib.error.URLError("Invalid url") from e
+
+    def _make_request(self, url=None):
+        """
+        Make github requests
+        """
+        if not url:
+            url = self.req_url
+        authorization = f'token {self.token}'
+        headers = {
+            "Accept": "application/vnd.github.v3+json",
+            "Authorization": authorization,
+        }
+        response = requests.get(url, headers=headers)
+        if response.status_code == 200:
+            return json.loads(response.text)
+        else:
+            if "API rate limit exceeded for" in response.text:
+                raise RateLimitException
```

Comment:
Is the response's status code not enough to determine this error? It would be much safer to check that than a string.

Reply:
Done